kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopCredentialsManager.scala

```diff
@@ -0,0 +1,225 @@
+   * @param send     Function to send encoded credentials to SQL engine
+   * @return `true` if credentials are sent successfully. Else, false.
+   * @throws KyuubiException if sending credentials fails.
+   */
+  def sendCredentialsIfNeeded(
+      engineId: String,
+      appUser: String,
+      send: String => Unit): Boolean = {
+    require(renewalExecutor != null, "renewalExecutor should be initialized")
+
+    val userRef = getUserRef(appUser)
+    val engineRef = getEngineRef(engineId)
+
+    var sent = false
+    if (userRef.getEpoch != engineRef.getEpoch) {
+      engineRef synchronized {
+        if (userRef.getEpoch != engineRef.getEpoch) {
+          val currentEpoch = userRef.getEpoch
+          val currentCreds = userRef.getEncodedCredentials
+          info(s"Send new credentials with epoch $currentEpoch to SQL engine $engineId")
+          Try(send(currentCreds)) match {
+            case Success(_) =>
+              info(s"Update SQL engine epoch from ${engineRef.getEpoch} to $currentEpoch")
+              engineRef.setEpoch(currentEpoch)
+              sent = true
+            case Failure(exception) =>
+              throw new KyuubiException(
+                s"Failed to send new credentials to SQL engine $engineId",
+                exception)
+          }
+        }
+      }
+    }
+    sent
+  }
+
+  // Visible for testing.
+  private[credentials] def getUserRef(appUser: String): UserCredentialsRef =
+    userRefMap.computeIfAbsent(
+      appUser,
+      appUser => {
+        val ref = new UserCredentialsRef(appUser)
+        scheduleRenewal(ref, 0)
+        info(s"Created CredentialsRef for user $appUser and scheduled a renewal task")
+        ref
+      })
+
+  // Visible for testing.
+  private[credentials] def getEngineRef(engineId: String): EngineCredentialsRef =
+    engineRefCache.getUnchecked(engineId)
+
+  // Visible for testing.
+  private[credentials] def isProviderLoaded(serviceName: String): Boolean =
+    providers.contains(serviceName)
+
+  // Visible for testing.
+  private[credentials] def isProviderRequired(serviceName: String): Boolean =
+    requiredProviders.contains(serviceName)
+
+  private def scheduleRenewal(userRef: UserCredentialsRef, delay: Long): Future[_] = {
+    val _delay = math.max(0, delay)
+    info(s"Scheduling renewal in ${Duration.ofMillis(_delay)}.")
+
+    val renewalTask = new Runnable {
+      override def run(): Unit = {
+        try {
+          val creds = new Credentials()
+          requiredProviders.values
+            .foreach(_.obtainDelegationTokens(hadoopConf, conf, userRef.getAppUser, creds))
+          userRef.updateCredentials(creds)
+          scheduleRenewal(userRef, renewalInterval)
+        } catch {
+          case _: InterruptedException =>
+          // Server is shutting down
+          case e: Exception =>
+            val duration = Duration.ofMillis(renewalRetryWait)
+            warn(s"Failed to update tokens, will try again in ${duration}", e)
```

Comment:
duration's string representation may hard to read

Reply:
Changed to represent milliseconds directly.